src/jit/codegenxarch.cpp

```diff
@@ -3848,21 +3849,73 @@ void CodeGen::genCodeForCpBlkRepMovs(GenTreeBlk* cpBlkNode)
     instGen(INS_r_movsb);
 }
 
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+#ifdef FEATURE_PUT_STRUCT_ARG_STK
+int CodeGen::genMove8IfNeeded(unsigned size, regNumber longTmpReg, GenTree* srcAddr, unsigned offset)
+{
+#ifdef _TARGET_X86_
+    instruction longMovIns = INS_movq;
+#else  // !_TARGET_X86_
+    instruction longMovIns = INS_mov;
+#endif // !_TARGET_X86_
+    if ((size & 8) != 0)
+    {
+        genCodeForLoadOffset(longMovIns, EA_8BYTE, longTmpReg, srcAddr, offset);
+        genStoreRegToStackArg(TYP_LONG, longTmpReg, offset);
+        return 8;
+    }
+    return 0;
+}
+
+int CodeGen::genMove4IfNeeded(unsigned size, regNumber intTmpReg, GenTree* srcAddr, unsigned offset)
+{
+    if ((size & 4) != 0)
+    {
+        genCodeForLoadOffset(INS_mov, EA_4BYTE, intTmpReg, srcAddr, offset);
+        genStoreRegToStackArg(TYP_INT, intTmpReg, offset);
+        return 4;
+    }
+    return 0;
+}
+
+int CodeGen::genMove2IfNeeded(unsigned size, regNumber intTmpReg, GenTree* srcAddr, unsigned offset)
+{
+    if ((size & 2) != 0)
+    {
+        genCodeForLoadOffset(INS_mov, EA_2BYTE, intTmpReg, srcAddr, offset);
+        genStoreRegToStackArg(TYP_SHORT, intTmpReg, offset);
+        return 2;
+    }
+    return 0;
+}
+
+int CodeGen::genMove1IfNeeded(unsigned size, regNumber intTmpReg, GenTree* srcAddr, unsigned offset)
+{
+
+    if ((size & 1) != 0)
+    {
+        genCodeForLoadOffset(INS_mov, EA_1BYTE, intTmpReg, srcAddr, offset);
+        genStoreRegToStackArg(TYP_BYTE, intTmpReg, offset);
+        return 1;
+    }
+    return 0;
+}
```

Comment:
Done.

---
In reply to: [79524614](https://github.com/dotnet/coreclr/pull/7252#discussion_r79524614) [](ancestors = 79524614)